### PR TITLE
Refactor NameTag Component

### PR DESCRIPTION
**Pull request by Shaomei**
Key changes include:

1. Unify all name tag input data into a react-hook-form
2. Get rid of the state management for nametag visibility
3. Simplify the state management for nametag content
4. Separate the drawing of handwave badge and nametag badge
5. Move the instances for ZoomApi and drawingApi outside App component so that they won't be instantiated everytime the app re-render.

Screenshot of current state:
Screenshot 2024-04-24 at 12 07 08 AM

### CODE REVIEW

**Comment by Shaomei** [comment]
FYI @kexinF @BrennanJones

File: app/main/state.ts
Comment (blickly) [inline app/main/state.ts:1]: I think if you search for 'nametag', you'll find there's more to remove from this file

**Review by blickly — COMMENTED** [review]
I think the tests might also need updating

File: app/main/page.tsx [hunk 1]
```diff
@@ -20,48 +22,56 @@ const apiList: Apis[] = [
   "removeVirtualForeground",
 ];
 
-function App() {
+const zoomConfigOptions: ConfigOptions = {
+  capabilities: apiList
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply app/main/page.tsx:26]: Done.

File: lib/draw_badge_api.ts [hunk 2]
```diff
@@ -26,6 +26,9 @@ export class DrawBadgeApi {
   constructor(private zoomApiWrapper: ZoomApiWrapper) {}
 
   private forceDrawing() {
+    //TODO: remove debugging code after finishing handWave component refactoring
+    console.log(this.nametag);
+    console.log(this.handwave);
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply lib/draw_badge_api.ts:31]: Done.

File: app/main/state.ts
Comment (Shaomei) [reply app/main/state.ts:1]: Removed.

File: app/main/page.tsx [hunk 3]
```diff
@@ -20,48 +22,56 @@ const apiList: Apis[] = [
...
-      capabilities: apiList
-    };
-    const zoomApiInstance: ZoomApiWrapper = createFromConfig(configOptions);
-
-    if (imageData) {
-      zoomApiInstance.setVirtualForeground(imageData);
-    } else {
-      zoomApiInstance.removeVirtualForeground();
-    }
+    console.log(state.selectedWaveHand);
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply app/main/page.tsx:61]: Removed.

File: components/NameTagForm.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  return (
+    <form onSubmit={handleSubmit(onNameTagContentChange)}>
+      <Controller
+        control={control}
+        name="visible"
+        defaultValue={false}
+        render={({ field: { onChange, value } }) => (
+          <FormControlLabel
+            control={
+              <Checkbox checked={value} onChange={onChange} />
```
[the remark this answers is not in the document]
Comment (Shaomei) [reply components/NameTagForm.tsx:40]: Done.

**Comment by Shaomei** [comment]
Passing all unit tests now.

```
> next-test-zoom-app@0.1.0 test
> jest

(node:54958) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
(Use `node --trace-deprecation ...` to show where the warning was created)
 PASS  __tests__/nametag.test.tsx
 PASS  __tests__/login.test.tsx
 PASS  __tests__/affirmation.test.tsx
 PASS  __tests__/mindfulness.test.tsx

Test Suites: 4 passed, 4 total
Tests:       10 passed, 10 total
Snapshots:   0 total
Time:        1.098 s
Ran all test suites.

```

File: app/main/page.tsx [hunk 5]
```diff
@@ -3,65 +3,68 @@
...
+  //TODO: store the new nametag content into DB
+  const updateNameTagContent: SubmitHandler<NameTagContent> = (data) => {
+    setNameTagContent(data);
+    foregroundDrawer.drawNameTag({
+      visible: data.visible,
+      fullName: data.fullName,
+      preferredName: data.preferredName,
+      pronouns: data.pronouns,
+      disclosure: data.disclosure,
+    });
```
Comment (blickly) [inline app/main/page.tsx:66]: TypeScript types are structural, so you can just do:
```suggestion
    foregroundDrawer.drawNameTag(data);
```

**Review by blickly — APPROVED** [review]

File: app/main/page.tsx (hunk 5 above)
Comment (Shaomei) [reply app/main/page.tsx:66]: Thanks! Done.

**Comment by Shaomei** [comment]
@BrennanJones: I just landed the NameTag refactoring, now you can start adding the DB write in updateNameTagContent function. You can also read saved nametag from DB (if any) and use it as the initial value for the NameTagForm. 

Note: In terms of the nameTag visibility checkbox, I don't think we should save it to the DB. It should be default to false when a user logs in, but maintain the checked/unchecked state when a user switch between tabs (not unchecked when re-rendering the tab). 

@kexinF I replace Switch with a Checkbox between it is a bit more robust in terms of the logic (Switch seems to be easy to be out of sync, but that perhaps could be fixed if we investigate it more later). I didn't put any effort into the CSS and I think some of the existing styles might have conflicted with the component styles (e.g. labelPlacement for the checkbox). Can you look into the CSS and help style the form better? Thank you!

**Pull request merged**